### PR TITLE
C175330 cleanup migration coord trunk

### PR DESCRIPTION
**Pull request by jonmeredith**
See [CASSANDRA-17533](https://issues.apache.org/jira/browse/CASSANDRA-17533)

### CODE REVIEW

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 1]
```diff
@@ -90,15 +93,20 @@
     private static final Logger logger = LoggerFactory.getLogger(MigrationCoordinator.class);
     private static final Future<Void> FINISHED_FUTURE = ImmediateFuture.success(null);
 
+    private static LongSupplier getUptimeFn = () -> ManagementFactory.getRuntimeMXBean().getUptime();
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java:97]: this is dead code in your patch

File: test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java [hunk 2]
```diff
@@ -228,7 +229,7 @@ public void accept(IInvokableInstance pullTo)
                 InetAddressAndPort endpoint = toCassandraInetAddressAndPort(pullFrom);
                 EndpointState state = Gossiper.instance.getEndpointStateForEndpoint(endpoint);
                 Gossiper.instance.doOnChangeNotifications(endpoint, ApplicationState.SCHEMA, state.getApplicationState(ApplicationState.SCHEMA));
-                Schema.instance.waitUntilReady(Duration.ofSeconds(10));
+                assertTrue("schema is ready", Schema.instance.waitUntilReady(Duration.ofSeconds(10)));
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java:232]: you are a good person here, but not in 4.0?  why the hate of 4.0!?!?!!

File: test/distributed/org/apache/cassandra/distributed/impl/Instance.java [hunk 3]
```diff
@@ -631,6 +633,7 @@ public void startup(ICluster cluster)
                 StorageService.instance.registerDaemon(CassandraDaemon.getInstanceForTesting());
                 if (config.has(GOSSIP))
                 {
+                    MigrationCoordinator.setUptimeFn(() -> TimeUnit.NANOSECONDS.toMillis(nanoTime() - startedAt));
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/impl/Instance.java]: spoke in slack, can you also fix the startedAt bug?  This is object init, which != uptime.

When we run for 1m and restart, the restart should have uptime=0 but will be uptime=1m due to the fact we don't cleanup the object.

**Review by dcapwell — CHANGES_REQUESTED** [review]

File: test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java (hunk 2 above)
Comment (jonmeredith) [reply test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java:232]: Puzzler. It was there, I'll put it back.

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 4]
```diff
@@ -467,16 +476,29 @@ void announce(UUID schemaVersion)
...
+            {
+                skipped = true;
+                return ImmediateFuture.success(null);
+            }
+            return executor.submit(task);
+        }
+        catch (RejectedExecutionException ex)
+        {
+            skipped = true;
             return ImmediateFuture.success(null);
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java:494]: shouldn't this return an failed future rather than success future?

Comment (dcapwell) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java:494]: 4.0 returned `null`.  The issue with this patch is that `executor.submit(task)` returns `Void`, which can only ever be `null`; so the caller has 0 way to know that the future was rejected; in 4.0 we could null check the future reference

Comment (dcapwell) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java:494]: removing my comment due to the fact this looks to be the desired behavior of this function, so you are not changing anything, you are just being consistent with existing logic...

**Review by dcapwell — APPROVED** [review]

**Comment by jonmeredith** [comment]
Reviewed, starting merge. Thanks for the review.

**Pull request closed without merging**